### PR TITLE
Translate Source Engine model's bone name exported from Crowbar to Japanese names from a defined list.

### PR DESCRIPTION
**Pull request by khanghugo**
The last thing needed is adding more base bones like 全ての親 and グルーブ but I am still trying to figure out how to use the API. Good API dude.

### CODE REVIEW

**Comment by khanghugo** [comment]
My PyCharm was acting crazy so the format here is not that good and lots of empty imports

**Comment by khanghugo** [comment]
I'm working to get the IK bones added.

File: python/translate_source_bone.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    # object
+    retme: pmxstruct.Pmx = pmxlib.read_pmx(input_filename_pmx, moreinfo=moreinfo)
+
+    for key in big_dict:
+        for index, name in enumerate(retme.bones):
+            if name.name_jp == key:
+                retme.bones[index].name_jp = big_dict[key]
+
+    output_filename_pmx = os.path.join(startpath, input_filename_pmx)
```
Comment (Nuthouse01) [inline python/translate_source_bone.py]: This bit will end badly
```suggestion
    output_filename_pmx = input_filename_pmx[0:-4] + "_sourcetrans.pmx"
```

**Comment by Nuthouse01** [comment]
Hey man, thanks for contributing! I've never had someone try to add to one of my Github repos before, so I'm kinda new to this whole deal. But I discovered that every time you add commits to this pull request, or add a comment, it sends me a notification. So please take a few days to tune and upgrade and debug things in your local repo, and when you're really sure you're done, then add all the new commits onto this pull request and add a comment saying "ok I'm done". Then I'll merge it in.
Seriously, no rush. Take your time, sleep on it, and be absolutely sure you're done before you tell me to merge it :-)

**Comment by khanghugo** [comment]
Sure, pretty much now I need to add some base bones and it should be done, some testing also because I did not test the code lol

**Comment by khanghugo** [comment]
It is ready to merge.

The core of the code is done. The problem is older Source Engine models are not working properly. Specifically, the some leg/waist weights are not in place/not belong to any bones. When we load VMD, there would be glitches from those parts. Newer models work well and flawlessly. If you want to check it, here they are: `https://drive.google.com/drive/folders/1_CfRiaQ-Xqd0rf-XrztCH6hnW1poSvuN?usp=sharing`. The only fixes for them is to draw the weights ourselves or just don't use older models. It could be fault from my part but I have no idea where.

The code does not work for all Crowbar's output since some more handcrafted models have different names rather than a set of generated frameworks.

Anyway, only the core is done, some other methods like `__name__` are not properly implemented if we compare to your scripts. It is also not that clean but it works and can be modified easily.

**Comment by khanghugo** [comment]
Meanwhile, I can look around to find a fix for the older models.

**Comment by khanghugo** [comment]
The final fix for that is not to do it. 

I think it is good to merge now. 
Still, I cannot understand how adding even empty bones would glitch the older models.

**Comment by khanghugo** [comment]
ok I'm done

Sorry for the spam earlier, I didn't have enough coffee.
Finally figured out what went wrong. The code is pretty scalable except for some hiccups but it works just fine.
Some methods I don't know quite sure if I should implement like `__name__` calling not only `main()` like you do. 

For ALL generic Source models, this script should work fine. Many models are not generic how they have different bone names.

I have the instructions and all the messages for printing if you want to use them for your executable add-ons.

File: python/nuthouse01_pmx_parser.py [hunk 2]
```diff
@@ -739,8 +739,10 @@ def encode_pmx_bones(nice: List[pmxstruct.PmxBone]) -> bytearray:
 		if bone.has_ik:  # ik:
 			# (ik_target, ik_loops, ik_anglelimit, ik_numlinks)
 			out += core.my_pack(fmt_bone_ik, [bone.ik_target_idx, bone.ik_numloops, bone.ik_angle, len(bone.ik_links)])
-			
-			for iklink in bone.ik_links:
+
+			for iklink_list in bone.ik_links:
+				# the normal iklink passes a list, not an object so the conditions below does not work
+				iklink = pmxstruct.PmxBoneIkLink(iklink_list[0], iklink_list[1], iklink_list[2])
```
Comment (Nuthouse01) [inline python/nuthouse01_pmx_parser.py:745]: revert this, I fixed the bug in your script instead

File: python/translate_source_bone.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    # if name == "ValveBiped.Bip01_R_Toe0":
+    #     retme.bones.insert(last_leg_item + 1, )
+
+    leg_left_ik_obj = pmxstruct.PmxBone(leg_left_ik_name, "", leg_left_ankle_pos, last_bone + 5, 0, False,
+                                        True, True, True, True, True,
+                                        False, [0.0, 0.0, 0.0], False, False, False,
+                                        False, False, None, None, None, None, None, None,
+                                        l_a_index, 40, 114.5916,
+                                        [[l_k_index, knee_limit_1, knee_limit_2],
+                                         [l_l_index, None, None]])
```
Comment (Nuthouse01) [inline python/translate_source_bone.py:333]: This is why you thought PMX parser was bugged, repeat for other 3 IK bones
```suggestion
                                        [pmxstruct.PmxBoneIkLink(idx=l_k_index, limit_min=knee_limit_1, limit_max=knee_limit_2), 
                                         pmxstruct.PmxBoneIkLink(idx=l_l_index)])
```

**Review by Nuthouse01 — CHANGES_REQUESTED** [review]
Your change in pmx_parser is incorrect, the bones you created were creating the IK links incorrectly. 
After you fix that, can you please cancel this pull request, download the latest version of the development branch, make your script work there, and then make a new pull request that tries to merge your changes with the development branch, instead of merging with master. 1) I don't want to modify the master branch until I've tried it for myself, 2) I created a new function "insert_single_bone" in file "_prune_unused_bones" which you should use to insert the IK bones, and can let you insert the motherbone/center/groove at the top of the bonelist instead of at the bottom.

**Comment by khanghugo** [comment]
Nice

**Pull request closed without merging**